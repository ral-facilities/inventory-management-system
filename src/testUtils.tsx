import { LocalizationProvider } from '@mui/x-date-pickers';
import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFnsV3';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { RenderOptions, render } from '@testing-library/react';
import { enGB } from 'date-fns/locale/en-GB';
import React from 'react';
import {
  RouterProvider,
  createBrowserRouter,
  createMemoryRouter,
} from 'react-router-dom';
import { paths } from './App';
import {
  CatalogueCategory,
  CatalogueCategoryProperty,
  CatalogueItem,
  Item,
  Manufacturer,
} from './api/api.types';
import CatalogueCategoriesJSON from './mocks/CatalogueCategories.json';
import CatalogueItemsJSON from './mocks/CatalogueItems.json';
import ItemsJSON from './mocks/Items.json';
import ManufacturersJSON from './mocks/Manufacturers.json';

export const createTestQueryClient = (): QueryClient =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        staleTime: 300000,
      },
    },
  });

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  queryClient?: QueryClient;
}

function constructRouterProvider(
  ui: React.ReactNode,
  queryClient: QueryClient,
  urlPathKey?: keyof typeof paths,
  initialEntry?: string
) {
  const Root: React.FunctionComponent = () => {
    return (
      <LocalizationProvider adapterLocale={enGB} dateAdapter={AdapterDateFns}>
        <QueryClientProvider client={queryClient}>{ui}</QueryClientProvider>
      </LocalizationProvider>
    );
  };

  const router =
    initialEntry !== undefined && urlPathKey !== undefined
      ? createMemoryRouter([{ path: paths[urlPathKey], Component: Root }], {
          initialEntries: [initialEntry],
        })
      : createBrowserRouter([{ path: '*', Component: Root }]);

  return { router: router, provider: <RouterProvider router={router} /> };
}

function constructRouterProviderWrapper(
  queryClient: QueryClient,
  urlPathKey?: keyof typeof paths,
  initialEntry?: string
) {
  const wrapper = ({
    children,
  }: {
    children: React.ReactNode;
  }): JSX.Element => {
    return constructRouterProvider(
      children,
      queryClient,
      urlPathKey,
      initialEntry
    ).provider;
  };
  return wrapper;
}

export function renderComponentWithRouterProvider(
  ui: React.ReactElement,
  urlPathKey?: keyof typeof paths,
  initialEntry?: string,
  {
    // Automatically create a query client instance if no query client was passed in
    queryClient = createTestQueryClient(),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  const { router, provider } = constructRouterProvider(
    ui,
    queryClient,
    urlPathKey,
    initialEntry
  );
  return {
    queryClient,
    router,
    ...render(ui, {
      wrapper: () => provider,
      ...renderOptions,
    }),
  };
}

export const hooksWrapperWithProviders = (props?: {
  queryClient?: QueryClient;
  urlPathKey?: keyof typeof paths;
  initialEntry?: string;
}) => {
  const testQueryClient = props?.queryClient ?? createTestQueryClient();
  return constructRouterProviderWrapper(
    testQueryClient,
    props?.urlPathKey,
    props?.initialEntry
  );
};

export const getCatalogueItemsPropertiesById = (
  id: string
): CatalogueCategoryProperty[] => {
  const filteredCategories = CatalogueCategoriesJSON.filter(
    (catalogueCategory) => catalogueCategory.id === id
  );

  if (filteredCategories.length === 0) {
    return [];
  }

  const properties = filteredCategories[0].properties ?? [];
  return properties as CatalogueCategoryProperty[];
};

export const getCatalogueCategoryById = (
  id: string
): CatalogueCategory | undefined => {
  return (
    (CatalogueCategoriesJSON.find(
      (catalogueCategory) => catalogueCategory.id === id
    ) as CatalogueCategory) || undefined
  );
};

export const getCatalogueItemById = (id: string): CatalogueItem | undefined => {
  return (
    (CatalogueItemsJSON.find(
      (catalogueItem) => catalogueItem.id === id
    ) as CatalogueItem) || undefined
  );
};

export const catalogueItemData = (id: string): CatalogueItem[] => {
  return CatalogueItemsJSON.filter(
    (catalogueItem) => catalogueItem.catalogue_category_id === id
  );
};

export const getManufacturerById = (id: string): Manufacturer | undefined => {
  return (
    (ManufacturersJSON.find(
      (manufacturer) => manufacturer.id === id
    ) as Manufacturer) || undefined
  );
};

export const getItemsByCatalogueItemId = (catalogueItemId: string): Item[] => {
  return ItemsJSON.filter((item) => item.catalogue_item_id === catalogueItemId);
};

export const getItemById = (itemId: string): Item | undefined => {
  return ItemsJSON.find((item) => item.id === itemId);
};

export const getItemsBySystemId = (systemId: string): Item[] => {
  return ItemsJSON.filter((item) => item.system_id === systemId);
};

export const CREATED_MODIFIED_TIME_VALUES = {
  created_time: '2024-01-01T12:00:00.000+00:00',
  modified_time: '2024-01-02T13:10:10.000+00:00',
};
