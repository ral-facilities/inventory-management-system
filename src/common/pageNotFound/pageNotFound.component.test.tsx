import { act } from 'react';
import {
  RoutesHomeLocation,
  type RoutesHomeLocationType,
} from '../../app.types';
import { renderComponentWithRouterProvider } from '../../testUtils';
import PageNotFoundComponent, {
  type PageNotFoundComponentProps,
} from './pageNotFound.component';

describe('Page Not Found Component', () => {
  let props: PageNotFoundComponentProps;

  const createView = () => {
    return renderComponentWithRouterProvider(
      <PageNotFoundComponent {...props} />
    );
  };

  it('renders the basic 404 page', async () => {
    let baseElement;
    await act(async () => {
      baseElement = createView().baseElement;
    });
    expect(baseElement).toMatchSnapshot();
  });

  Object.keys(RoutesHomeLocation).forEach((route) => {
    it(`renders ${route} 404 page`, async () => {
      props = { homeLocation: route as RoutesHomeLocationType };
      let baseElement;
      await act(async () => {
        baseElement = createView().baseElement;
      });
      expect(baseElement).toMatchSnapshot();
    });
  });
});
