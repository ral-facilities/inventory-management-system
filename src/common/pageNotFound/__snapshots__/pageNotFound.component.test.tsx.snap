// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Page Not Found Component > renders Admin 404 page 1`] = `
<body>
  <div>
    <div>
      <div
        class="MuiBox-root css-d8r78j"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1wffj3i-MuiSvgIcon-root"
          data-testid="BugReportIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 8h-2.81c-.45-.78-1.07-1.45-1.82-1.96L17 4.41 15.59 3l-2.17 2.17C12.96 5.06 12.49 5 12 5c-.49 0-.96.06-1.41.17L8.41 3 7 4.41l1.62 1.63C7.88 6.55 7.26 7.22 6.81 8H4v2h2.09c-.05.33-.09.66-.09 1v1H4v2h2v1c0 .34.04.67.09 1H4v2h2.81c1.04 1.79 2.97 3 5.19 3s4.15-1.21 5.19-3H20v-2h-2.09c.05-.33.09-.66.09-1v-1h2v-2h-2v-1c0-.34-.04-.67-.09-1H20zm-6 8h-4v-2h4zm0-4h-4v-2h4z"
          />
        </svg>
        <p
          class="MuiTypography-root MuiTypography-body1 css-1fjbu-MuiTypography-root"
        >
          404
        </p>
      </div>
      <div
        class="MuiBox-root css-sg7ujc"
      >
        <h2
          class="MuiTypography-root MuiTypography-h2 css-18ft2kn-MuiTypography-root"
        >
          Page not found
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body1 css-n3u125-MuiTypography-root"
        >
          We're sorry, the page you requested was not found on the server. If you entered the URL manually please check your spelling and try again. Otherwise, return to the
           
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-17fxp5y-MuiTypography-root-MuiLink-root"
            href="/admin-ims"
          >
            admin
             home page
          </a>
           
          .
        </p>
      </div>
    </div>
  </div>
</body>
`;

exports[`Page Not Found Component > renders Catalogue 404 page 1`] = `
<body>
  <div>
    <div>
      <div
        class="MuiBox-root css-d8r78j"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1wffj3i-MuiSvgIcon-root"
          data-testid="BugReportIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 8h-2.81c-.45-.78-1.07-1.45-1.82-1.96L17 4.41 15.59 3l-2.17 2.17C12.96 5.06 12.49 5 12 5c-.49 0-.96.06-1.41.17L8.41 3 7 4.41l1.62 1.63C7.88 6.55 7.26 7.22 6.81 8H4v2h2.09c-.05.33-.09.66-.09 1v1H4v2h2v1c0 .34.04.67.09 1H4v2h2.81c1.04 1.79 2.97 3 5.19 3s4.15-1.21 5.19-3H20v-2h-2.09c.05-.33.09-.66.09-1v-1h2v-2h-2v-1c0-.34-.04-.67-.09-1H20zm-6 8h-4v-2h4zm0-4h-4v-2h4z"
          />
        </svg>
        <p
          class="MuiTypography-root MuiTypography-body1 css-1fjbu-MuiTypography-root"
        >
          404
        </p>
      </div>
      <div
        class="MuiBox-root css-sg7ujc"
      >
        <h2
          class="MuiTypography-root MuiTypography-h2 css-18ft2kn-MuiTypography-root"
        >
          Page not found
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body1 css-n3u125-MuiTypography-root"
        >
          We're sorry, the page you requested was not found on the server. If you entered the URL manually please check your spelling and try again. Otherwise, return to the
           
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-17fxp5y-MuiTypography-root-MuiLink-root"
            href="/catalogue"
          >
            catalogue
             home page
          </a>
           
          .
        </p>
      </div>
    </div>
  </div>
</body>
`;

exports[`Page Not Found Component > renders Manufacturers 404 page 1`] = `
<body>
  <div>
    <div>
      <div
        class="MuiBox-root css-d8r78j"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1wffj3i-MuiSvgIcon-root"
          data-testid="BugReportIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 8h-2.81c-.45-.78-1.07-1.45-1.82-1.96L17 4.41 15.59 3l-2.17 2.17C12.96 5.06 12.49 5 12 5c-.49 0-.96.06-1.41.17L8.41 3 7 4.41l1.62 1.63C7.88 6.55 7.26 7.22 6.81 8H4v2h2.09c-.05.33-.09.66-.09 1v1H4v2h2v1c0 .34.04.67.09 1H4v2h2.81c1.04 1.79 2.97 3 5.19 3s4.15-1.21 5.19-3H20v-2h-2.09c.05-.33.09-.66.09-1v-1h2v-2h-2v-1c0-.34-.04-.67-.09-1H20zm-6 8h-4v-2h4zm0-4h-4v-2h4z"
          />
        </svg>
        <p
          class="MuiTypography-root MuiTypography-body1 css-1fjbu-MuiTypography-root"
        >
          404
        </p>
      </div>
      <div
        class="MuiBox-root css-sg7ujc"
      >
        <h2
          class="MuiTypography-root MuiTypography-h2 css-18ft2kn-MuiTypography-root"
        >
          Page not found
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body1 css-n3u125-MuiTypography-root"
        >
          We're sorry, the page you requested was not found on the server. If you entered the URL manually please check your spelling and try again. Otherwise, return to the
           
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-17fxp5y-MuiTypography-root-MuiLink-root"
            href="/manufacturers"
          >
            manufacturers
             home page
          </a>
           
          .
        </p>
      </div>
    </div>
  </div>
</body>
`;

exports[`Page Not Found Component > renders Systems 404 page 1`] = `
<body>
  <div>
    <div>
      <div
        class="MuiBox-root css-d8r78j"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1wffj3i-MuiSvgIcon-root"
          data-testid="BugReportIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 8h-2.81c-.45-.78-1.07-1.45-1.82-1.96L17 4.41 15.59 3l-2.17 2.17C12.96 5.06 12.49 5 12 5c-.49 0-.96.06-1.41.17L8.41 3 7 4.41l1.62 1.63C7.88 6.55 7.26 7.22 6.81 8H4v2h2.09c-.05.33-.09.66-.09 1v1H4v2h2v1c0 .34.04.67.09 1H4v2h2.81c1.04 1.79 2.97 3 5.19 3s4.15-1.21 5.19-3H20v-2h-2.09c.05-.33.09-.66.09-1v-1h2v-2h-2v-1c0-.34-.04-.67-.09-1H20zm-6 8h-4v-2h4zm0-4h-4v-2h4z"
          />
        </svg>
        <p
          class="MuiTypography-root MuiTypography-body1 css-1fjbu-MuiTypography-root"
        >
          404
        </p>
      </div>
      <div
        class="MuiBox-root css-sg7ujc"
      >
        <h2
          class="MuiTypography-root MuiTypography-h2 css-18ft2kn-MuiTypography-root"
        >
          Page not found
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body1 css-n3u125-MuiTypography-root"
        >
          We're sorry, the page you requested was not found on the server. If you entered the URL manually please check your spelling and try again. Otherwise, return to the
           
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-17fxp5y-MuiTypography-root-MuiLink-root"
            href="/systems"
          >
            systems
             home page
          </a>
           
          .
        </p>
      </div>
    </div>
  </div>
</body>
`;

exports[`Page Not Found Component > renders the basic 404 page 1`] = `
<body>
  <div>
    <div>
      <div
        class="MuiBox-root css-d8r78j"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1wffj3i-MuiSvgIcon-root"
          data-testid="BugReportIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 8h-2.81c-.45-.78-1.07-1.45-1.82-1.96L17 4.41 15.59 3l-2.17 2.17C12.96 5.06 12.49 5 12 5c-.49 0-.96.06-1.41.17L8.41 3 7 4.41l1.62 1.63C7.88 6.55 7.26 7.22 6.81 8H4v2h2.09c-.05.33-.09.66-.09 1v1H4v2h2v1c0 .34.04.67.09 1H4v2h2.81c1.04 1.79 2.97 3 5.19 3s4.15-1.21 5.19-3H20v-2h-2.09c.05-.33.09-.66.09-1v-1h2v-2h-2v-1c0-.34-.04-.67-.09-1H20zm-6 8h-4v-2h4zm0-4h-4v-2h4z"
          />
        </svg>
        <p
          class="MuiTypography-root MuiTypography-body1 css-1fjbu-MuiTypography-root"
        >
          404
        </p>
      </div>
      <div
        class="MuiBox-root css-sg7ujc"
      >
        <h2
          class="MuiTypography-root MuiTypography-h2 css-18ft2kn-MuiTypography-root"
        >
          Page not found
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body1 css-n3u125-MuiTypography-root"
        >
          We're sorry, the page you requested was not found on the server. If you entered the URL manually please check your spelling and try again. Otherwise, return to the
           
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-17fxp5y-MuiTypography-root-MuiLink-root"
            href="/"
          >
             home page
          </a>
           
          .
        </p>
      </div>
    </div>
  </div>
</body>
`;
